src/QAT/qat/purr/backends/echo.py

```diff
@@ -52,32 +55,98 @@ def apply_setup_to_hardware(hw, qubit_count: int = 4):
+            for other_qubit in qubit_devices:
+                if qubit != other_qubit:
+                    qubit.create_pulse_channel(
+                        auxiliary_devices=[other_qubit],
+                        channel_type=ChannelType.cross_resonance,
+                        frequency=5.5e9,
+                        scale=50,
+                    )
+                    qubit.create_pulse_channel(
+                        auxiliary_devices=[other_qubit],
+                        channel_type=ChannelType.cross_resonance_cancellation,
+                        frequency=5.5e9,
+                        scale=0.0,
+                    )
+                qubit.add_coupled_qubit(qubit_devices[(i + 1) % qubit_count])
+                qubit.add_coupled_qubit(qubit_devices[(i - 1) % qubit_count])
+    else:
+        for connection in connectivity:
+            left_index, right_index = connection
+            qubit_left = qubits_by_index.get(left_index)
+            qubit_right = qubits_by_index.get(right_index)
+            qubit_right.create_pulse_channel(
+                auxiliary_devices=[qubit_left],
+                channel_type=ChannelType.cross_resonance,
+                frequency=5.5e9,
+                scale=50,
+            )
+            qubit_right.create_pulse_channel(
+                auxiliary_devices=[qubit_left],
+                channel_type=ChannelType.cross_resonance_cancellation,
+                frequency=5.5e9,
+                scale=0.0,
+            )
+            qubit_left.create_pulse_channel(
+                auxiliary_devices=[qubit_right],
+                channel_type=ChannelType.cross_resonance,
+                frequency=5.5e9,
+                scale=50,
+            )
+            qubit_left.create_pulse_channel(
+                auxiliary_devices=[qubit_right],
+                channel_type=ChannelType.cross_resonance_cancellation,
+                frequency=5.5e9,
+                scale=0.0,
+            )
+            qubit_left.add_coupled_qubit(qubit_right)
+            qubit_right.add_coupled_qubit(qubit_left)
 
     hw.add_quantum_device(*qubit_devices, *resonator_devices)
     hw.is_calibrated = True
+
     return hw
 
 
-def get_default_echo_hardware(qubit_count=4) -> "QuantumHardwareModel":
+def generate_connectivity(con_type, qubit_count):
+    if con_type == Connectivity.Ring:
+        if qubit_count == 1:
+            return []
+        if qubit_count == 2:
+            return [(0, 1)]
+        return [(i, (i + 1) % qubit_count) for i in range(qubit_count)]
+    return None
+
+
+def add_direction_couplings_to_hardware(model, connectivity):
+    for edge in connectivity:
+        model.qubit_direction_couplings.append(
+            QubitCoupling(edge, quality=random.randrange(1, 20))
+        )
+    return model
+
+
+class Connectivity(Enum):
+    Ring = auto()
+
+
+def get_default_echo_hardware(
+    qubit_count=4,
+    connectivity: Optional[Union[Connectivity, List[Tuple[int, int]]]] = None,
```

Comment:
Default `connectivity` set to `None` in order to retain backwards compatibility.